java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -491,11 +492,46 @@ public void visitLambdaExpression(LambdaExpressionTree tree) {
         }
       }
       super.visitLambdaExpression(tree);
+      if(lambdaType.isUnknown() || lambdaType.isTagged(JavaType.DEFERRED)) {
+        return;
+      }
+      refineLambdaType(lambdaExpressionTree, lambdaType);
     } else {
       registerType(tree, symbols.deferedType(lambdaExpressionTree));
     }
   }
 
+  private void refineLambdaType(LambdaExpressionTreeImpl lambdaExpressionTree, JavaType lambdaType) {
+    JavaType samReturnType = (JavaType) getSamMethod(lambdaType).returnType().type();
+    JavaType capturedReturnType = resolve.resolveTypeSubstitution(samReturnType, lambdaType);
+    if (lambdaExpressionTree.body().is(Tree.Kind.BLOCK) || capturedReturnType.is("void")) {
+      // TODO compute return type for lambda with block and return statements
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)